scripts/content/vacms-5744/library.php

```diff
@@ -0,0 +1,746 @@
+    }
+  }
+  return $result;
+}
+
+/**
+ * Compare relevant info from two associative arrays built from nodes.
+ *
+ * @param array $array1
+ *   The first array.
+ * @param array $array2
+ *   The second array.
+ *
+ * @return array
+ *   An array composed of the differences between the two arrays.
+ */
+function compare_filtered_arrays(array $array1, array $array2): array {
+  return compare_arrays(filter_array($array1), filter_array($array2));
+}
+
+/**
+ * Retrieve a list of all paragraph item fields on all entities.
+ *
+ * We can do this instead of retrieving the field config because we're only
+ * worried about fields with existing data.
+ *
+ * @param string $parent_type
+ *   The parent entity type.
+ *
+ * @return string[]
+ *   A list of field names with existing paragraph items.
+ */
+function get_paragraph_field_names(string $parent_type = 'node'): array {
+  $database = \Drupal::database();
+
+  $query = $database->select('paragraphs_item_field_data', 'p');
+  $query->condition('parent_type', $parent_type);
+  $query->addField('p', 'parent_field_name', 'parent_field_name');
+  $result = $query->distinct()->execute()->fetchCol();
+
+  return $result;
+}
+
+/**
+ * Retrieve a list of paragraph IDs and the parents to which they point.
+ *
+ * This queries for distinct (pid, nid) tuples, and then filters this
+ * paragraphs with revisions referencing multiple parent nodes.
+ *
+ * @param string $parent_type
+ *   The parent type, if desired.
+ *
+ * @return array[]
+ *   An associative array with the form $pid => [ $nid1, $nid2, ...].
+ */
+function get_multiply_parented_paragraph_hash(string $parent_type = NULL): array {
+  $database = \Drupal::database();
+
+  $query = $database->select("paragraphs_item_revision_field_data", 'paragraphs_item_revision');
+  $query->join('node_revision', 'node', "node.nid=paragraphs_item_revision.parent_id");
+  $query->addField('paragraphs_item_revision', 'id', 'paragraph_id');
+  $query->addField('paragraphs_item_revision', 'parent_id');
+  $query->condition('parent_id', "23", "<>");
+  $query->condition('parent_id', "0", "<>");
+  if (isset($parent_type)) {
+    $query->condition('parent_type', $parent_type);
+  }
+  $statement = $query->distinct()->execute();
+
+  $hash = [];
+  foreach ($statement as $item) {
+    $pid = (int) $item->paragraph_id;
+    $nid = (int) $item->parent_id;
+    $hash[$pid] = $hash[$pid] ?? [];
+    $hash[$pid][] = $nid;
+  }
+
+  $result = array_filter($hash, function ($nid_list) {
+    return count($nid_list) > 1;
+  });
+
+  ksort($result);
+
+  return $result;
+}
+
+/**
+ * Retrieve a list of all parents of multiply-parented paragraphs.
+ *
+ * This is an inversion of get_multiply_parented_paragraph_hash().
+ *
+ * @return array[]
+ *   An associative array with the form $nid => [ $pid1, $pid2, ...].
+ */
+function get_multiply_parented_paragraph_parent_hash(string $parent_type = NULL): array {
+  $pidHash = get_multiply_parented_paragraph_hash($parent_type);
```

Comment:
```suggestion
  $pid_hash = get_multiply_parented_paragraph_hash($parent_type);
```

just for consistent case...